xwiki-platform-core/xwiki-platform-oldcore/src/main/java/org/xwiki/store/hibernate/HibernateAdapter.java

```diff
@@ -0,0 +1,140 @@
+
+/**
+ * Implement XWiki Hibernate features specific to a RDBMS. The goal is mainly to extend the Hibernate {@link Dialect}
+ * and JDBC {@link Driver} concepts with features they don't support (table compression, wiki -> database name, etc.).
+ * <p>
+ * It's recommended to extend {@link AbstractHibernateAdapter}.
+ * <p>
+ * A {@link HibernateAdapter} is supposed to be provided by a {@link HibernateAdapterFactory}, but if you need to
+ * associate the {@link HibernateAdapter} to a specific database engine (and optionally to a specific minimum version)
+ * you can use the following helper format:
+ * <ul>
+ * <li>"jdbc scheme/minimum version" (for example the "mysql/8" adapter will be selected for a MySQL 9.0.1 server, if no
+ * adapter is registered for version 9)</li>
+ * <li>"jdbc scheme" (for example the "mysql" adapter will be selected for a MySQL 5.7 server, if no adapter is
+ * registered for version 5.7 or lower)</li>
+ * </ul>
+ * 
+ * @version $Id$
+ * @since 17.1.0RC1
+ */
+@Unstable
+@Role
+public interface HibernateAdapter
+{
+    /**
+     * The name of the meta attribute indicating if a table should be compressed.
+     */
+    String META_ATTRIBUTE_COMPRESSED = "xwiki-compressed";
+
+    // Configuration
+
+    /**
+     * Get the native database/schema name for the current wiki.
+     *
+     * @return the native database/schema name
+     */
+    String getDatabaseFromWikiName();
+
+    /**
+     * Get the native database/schema name for the passed wiki identifier.
+     *
+     * @param wikiId the identifier of the wiki
+     * @return the native database/schema name
+     */
+    String getDatabaseFromWikiName(String wikiId);
+
+    /**
+     * @return true if the user has configured Hibernate to use XWiki in schema mode (vs database mode)
+     */
+    boolean isConfiguredInSchemaMode();
+
+    /**
+     * @param persistentClass the Hibernate entity
+     * @return the table name
+     */
+    String getTableName(PersistentClass persistentClass);
+
+    /**
+     * @param table the Hibernate table representation
+     * @return the table name
+     */
+    String getTableName(Table table);
+
+    /**
+     * @param tableName the name of the table
+     * @return the name of the table in the right case/format
+     */
+    String getTableName(String tableName);
+
+    /**
+     * @param entity the Hibernate entity for which to extract the configuration
+     * @return true if the table should be compressed
+     */
+    boolean isCompressed(PersistentClass entity);
+
+    /**
+     * Escape database name to be used in a query.
+     *
+     * @param databaseName the schema name to escape
+     * @return the escaped version
+     */
+    String escapeDatabaseName(String databaseName);
+
+    // Global
+
+    /**
+     * @return the Hibernate {@link SessionFactory}
+     */
+    SessionFactory getSessionFactory();
+
+    /**
+     * @return the Hibernate {@link Dialect}
+     */
+    Dialect getDialect();
```

Comment:
I checked quickly and apparently JdbcServicesImpl could return null for the dialect, you sure it can never happen here?

Reply:
It's impossible to have an adapter in the first place without a connection to the database (since the choice of adapter depends on metadata gathered from the server), and if there is a Hibernate connection there is a dialect.